### PR TITLE
DDCNL-2222 | Play 2.6 upgrade and refactor

### PR DESCRIPTION
**Pull request by liampan**
scala fmt has been added but not yet run, this will be done in a future PR.

### CODE REVIEW

File: test/services/CitizenIncomeServiceSpec.scala [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+        "AA555555A" -> errorNotFound,
+        "AA666666A" -> errorNotFoundNino,
+        "an unmatched Nino" -> null
+      ),
+      500 -> List(
+        "AA777777A" -> serverError,
+        "AA888888A" -> serviceUnavailable
+      )
+    ) foreach {
+      case (expectedStatus: Int, l: List[(String, JsValue)]) =>
```
Comment (Tateni93) [inline test/services/CitizenIncomeServiceSpec.scala:55]: Can we get a better name for the second value

File: app/repositories/CitizenIncomeRepository.scala [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+  val successMatchTwoTaxYears: JsValue = getJsValue("/resources/200-success-matched-two-tax-years.json")
+  val successNoMatch: JsValue = getJsValue("/resources/200-success-no-match.json")
+  val singleEmpSingleTaxYear: JsValue = getJsValue("/resources/AA777771A_rti.json")
+  val multipleEmpSingleTaxYear: JsValue = getJsValue("/resources/AA777772A_rti.json")
+  val multipleEmpMultipleTaxYears: JsValue = getJsValue("/resources/AA777773A_rti.json")
+  val multipleEmpMultipleTaxYearsOp: JsValue = getJsValue("/resources/AA777774A_rti.json")
+  val multipleEmpMultipleTaxYearsYdr: JsValue = getJsValue("/resources/AA777775A_rti.json")
+  val validNinoWithNoData: JsValue = getJsValue("/resources/AA777776A_rti.json")
+
+
```
Comment (Dandyboyo2) [inline app/repositories/CitizenIncomeRepository.scala:40]: remove one empty line

File: test/controllers/RetrieveCitizenIncomeControllerSpec.scala [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+  def fakeRequest(json: JsValue = validRequestJson): FakeRequest[JsValue] = FakeRequest("POST", "/", FakeHeaders(), json)
+    .withHeaders(("Content-type", "application/json"))
+
+  val mockStubService: CitizenIncomeService = mock[CitizenIncomeService]
+  val mockSchemaValidation: SchemaValidation = mock[SchemaValidation]
+  val SUT: RetrieveCitizenIncomeController = inject[RetrieveCitizenIncomeController]
+
+  override def beforeEach(): Unit = {
+    super.beforeEach()
+    reset(mockStubService)
```
Comment (danewatts) [inline test/controllers/RetrieveCitizenIncomeControllerSpec.scala:67]: Need to reset the SchemaValidation mock

File: test/services/SchemaValidationSpec.scala [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package services
+
+import org.scalatestplus.play.PlaySpec
+import play.api.libs.json._
+
+class SchemaValidationSpec extends PlaySpec {
```
Comment (danewatts) [inline test/services/SchemaValidationSpec.scala:22]: Few extra spaces in here

File: app/controllers/RetrieveCitizenIncomeController.scala [hunk 5]
```diff
@@ -16,86 +16,25 @@
...
+                                                 cc: ControllerComponents
+                                               ) extends BackendController(cc) {
+
+  def getRetrieveCitizenIncome(nino: String): Action[JsValue] = Action(parse.json) { implicit request =>
+
+    if(schema.isJsonValid(request.body)){
+      citizenIncomeService.getRetrieveCitizenIncome(nino)
+    } else {
+      BadRequest(
+        Json.parse("""{"code":"INVALID_PAYLOAD","reason":"Submission has not passed validation. Invalid Payload."}"""))
```
Comment (liampan) [inline app/controllers/RetrieveCitizenIncomeController.scala:37]: why do we json parse this? its a common pattern is there a good reason?

Comment (liampan) [reply app/controllers/RetrieveCitizenIncomeController.scala:37]: ^ as apposed to defining a JsObject or just using the string, surely this has potential to throw runtime exceptions 

potentially: 
JsObject(Map("code" -> JsString("INVALID_PAYLOAD"), "reason" -> JsString("Submission has not passed validation. Invalid Payload.")))

File: test/controllers/RetrieveCitizenIncomeControllerSpec.scala [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+            |  "surname": "Smith",
+            |  "firstName": "John",
+            |  "gender": "M"
+            |}""".stripMargin
+        )
+
+        val invalidJsonFakeRequest = fakeRequest(invalidRequestJson)
+        when(mockSchemaValidation.isJsonValid(ArgumentMatchers.any())).thenReturn(false)
+
+        val response = SUT.getRetrieveCitizenIncome("AA111111A")(invalidJsonFakeRequest)
```
[the remark this answers is not in the document]
Comment (thirlz95) [reply test/controllers/RetrieveCitizenIncomeControllerSpec.scala:121]: Sorted

File: test/services/CitizenIncomeServiceSpec.scala (hunk 1 above)
Comment (thirlz95) [reply test/services/CitizenIncomeServiceSpec.scala:55]: Changed to testCases naming convention

File: test/services/CitizenIncomeServiceSpec.scala [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+import repositories.CitizenIncomeRepository._
+
+import scala.concurrent.Future
+
+class CitizenIncomeServiceSpec extends PlaySpec {
+
+  val SUT = new CitizenIncomeService
+
+  "getRetrieveCitizenIncome" must {
+    Map(
```
[the remark this answers is not in the document]
Comment (thirlz95) [reply test/services/CitizenIncomeServiceSpec.scala:32]: Changed to Seq

File: app/repositories/CitizenIncomeRepository.scala (hunk 2 above)
Comment (thirlz95) [reply app/repositories/CitizenIncomeRepository.scala:40]: Removed line

File: test/controllers/RetrieveCitizenIncomeControllerSpec.scala (hunk 3 above)
Comment (thirlz95) [reply test/controllers/RetrieveCitizenIncomeControllerSpec.scala:67]: added SchemaValidation to the reset function

File: test/services/SchemaValidationSpec.scala (hunk 4 above)
Comment (thirlz95) [reply test/services/SchemaValidationSpec.scala:22]: done

File: app/services/CitizenIncomeService.scala [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+    "AA888888A" -> InternalServerError(serviceUnavailable),
+    "AA777771A" -> Ok(singleEmpSingleTaxYear),
+    "AA777772A" -> Ok(multipleEmpSingleTaxYear),
+    "AA777773A" -> Ok(multipleEmpMultipleTaxYears),
+    "AA777774A" -> Ok(multipleEmpMultipleTaxYearsOp),
+    "AA777775A" -> Ok(multipleEmpMultipleTaxYearsYdr),
+    "AA777776A" -> Ok(validNinoWithNoData)
+  ).withDefault( nino => {
+    logger.debug(s"Nino $nino was not found.")
+    NotFound
```
[the remark this answers is not in the document]
Comment (thirlz95) [reply app/services/CitizenIncomeService.scala:46]: added "The Resource that was not requested was not found."

File: test/controllers/RetrieveCitizenIncomeControllerSpec.scala [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+      |  "toDate": "2017-12-31",
+      |  "surname": "Smith",
+      |  "firstName": "John",
+      |  "gender": "M",
+      |  "initials": "J B",
+      |  "dateOfBirth": "2000-03-29"
+      |}""".stripMargin
+  )
+
+  def fakeRequest(json: JsValue = validRequestJson): FakeRequest[JsValue] = FakeRequest("POST", "/", FakeHeaders(), json)
```
[the remark this answers is not in the document]
Comment (Tateni93) [reply test/controllers/RetrieveCitizenIncomeControllerSpec.scala:58]: Done

File: test/controllers/RetrieveCitizenIncomeControllerSpec.scala [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+import play.api.test.{FakeHeaders, FakeRequest, Injecting, NoMaterializer}
+import services.{CitizenIncomeService, SchemaValidation}
+
+import scala.concurrent.Future
+
+class RetrieveCitizenIncomeControllerSpec extends PlaySpec with GuiceOneAppPerSuite with BeforeAndAfterEach with MockitoSugar with Injecting with ScalaCheckPropertyChecks {
+
+  implicit val noMaterializer: Materializer = NoMaterializer
+
+  val nino: String = "AA111111A"
```
[the remark this answers is not in the document]
Comment (Tateni93) [reply test/controllers/RetrieveCitizenIncomeControllerSpec.scala:44]: Changed

File: build.sbt [hunk 11]
```diff
@@ -0,0 +1,39 @@
+import play.sbt.routes.RoutesKeys
+import uk.gov.hmrc.sbtdistributables.SbtDistributablesPlugin.publishingSettings
+import scoverage.ScoverageKeys
+
+
+val appName = "retrieve-citizen-income-stub"
+
+lazy val scoverageSettings = {
+  val sCoverageExcludesPattens = List(
```
[the remark this answers is not in the document]
Comment (Tateni93) [reply build.sbt:9]: Changed

File: app/controllers/RetrieveCitizenIncomeController.scala (hunk 5 above)
Comment (Tateni93) [reply app/controllers/RetrieveCitizenIncomeController.scala:37]: This has been amended

[the remark this answers is not in the document]
Comment (Tateni93) [reply app/controllers/RetrieveCitizenIncomeController.scala:37]: Already got it

File: app/services/CitizenIncomeService.scala (hunk 8 above)
[the remark this answers is not in the document]
Comment (Tateni93) [reply app/services/CitizenIncomeService.scala]: Already done

File: app/services/SchemaValidation.scala [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+package services
+
+import com.eclipsesource.schema.{SchemaType, SchemaValidator}
+import com.google.inject.Inject
+import play.api.Logger
+import play.api.libs.json.{JsError, JsSuccess, JsValue, Json}
+
+import scala.io.Source
+
+class SchemaValidation @Inject()() {
```
[the remark this answers is not in the document]
Comment (Tateni93) [reply app/services/SchemaValidation.scala:26]: Already done

File: app/services/SchemaValidation.scala [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+
+  private val requestSchema: SchemaType = {
+    val resource = getClass.getResourceAsStream("/schemas/des-request-schema-v1.json")
+    val json = Json.parse(Source.fromInputStream(resource).mkString)
+    json.validate[SchemaType].getOrElse(throw new RuntimeException("Json Schema is not valid Json"))
+  }
+
+  private val validator: SchemaValidator = SchemaValidator()
+
+  def isJsonValid(json: JsValue): Boolean = {
```
[the remark this answers is not in the document]
Comment (Tateni93) [reply app/services/SchemaValidation.scala:38]: We have decided to leave this as it is.

**Pull request merged**